File: Prefix/560.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        counts = defaultdict(int)
        counts[0] = 1
        ans = sumcurr = 0

        for num in nums:
            sumcurr += num
            # this is the key, since sumcurr -(sumcurr-k) = k
            ans += counts[sumcurr - k]
            counts[sumcurr] += 1
        return ans

File: Prefix/560.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> save = new HashMap<>();
        // by default, sumcurr - k = 0 initialized in hashmap
        save.put(0,1);
        int curr = 0;
        int ans = 0;

        for (int num: nums){
            curr += num;

            // curr_sum - (curr_sum-k) = k
            if (save.containsKey(curr-k)){
                ans += save.get(curr-k);
            }
            save.put(curr, save.getOrDefault(curr, 0)+1);
        }

        return ans;
    }
}